src/backend/distributed/commands/multi_copy.c

```diff
@@ -3578,6 +3600,18 @@ CopyGetPlacementConnection(HTAB *connectionStateHash, ShardPlacement *placement,
 	}
 
 	connection = GetPlacementConnection(connectionFlags, placement, nodeUser);
+	if (connection == NULL)
+	{
+		/*
+		 * The connection manager throttled any new connections, so pick an existing
+		 * connection with least utilization.
+		 */
+		connection =
+			GetLeastUtilisedCopyConnection(copyConnectionStateList, nodeName, nodePort);
+		Assert(connection != NULL);
+
+		return connection;
```

Comment:
I wonder if it's ok to return an existing query without checking if the connection is ok, as we do in the following lines:

```c
if (PQstatus(connection->pgConn) != CONNECTION_OK)
```

Reply:
As we early return in many cases now, we should properly use the following code block when necessary:
```C
	MarkRemoteTransactionCritical(connection);

	if (MultiShardConnectionType != SEQUENTIAL_CONNECTION)
	{
		ClaimConnectionExclusively(connection);
	}
```

We should call this if the connection is NOT returned from `GetLeastUtilisedCopyConnection`, meaning that we're using the connection for this COPY command for the first time.

Reply:
I run into issues when I tried to mark the connections that were used to access placements as exclusive.

You can see my attempt in this commit: 87f185ba102eb016109e2b6cfc0d87bf86b9c604

Here is one of the places where copy started to fail, after I marked it as exclusive:
https://github.com/citusdata/citus/blob/87f185ba102eb016109e2b6cfc0d87bf86b9c604/src/test/regress/expected/multi_modifying_xacts.out#L221-L227

Reply:
hmz, something is a little off here.

The error message is right and the bug is that we should not be claiming connections that still need to be used for other placements, but then is it ever ok to claim connections on the COPY logic? (i.e. do we have a bug even without the extra call?)

In the adaptive executor, we first assign tasks to connections, and only them claim the connections exclusively. We know that after this, any subsequent task does not have an assigned connection, so we never reach a point where we need to use a specific connection when that connection is claimed exclusively.

Reply:
I guess CONNECTION_PER_PLACEMENT saved us in the past by ensuring we only claim clean connections, so we never ran into a situation where GetConnectionIfPlacementAccessedInXact errors out later. Maybe revert that commit?

Reply:
I guess what we would really want is some variant of "claimed exclusively" which instead of "return this connection to no one" means "return this connection only to me".